eth-rpc/client/clientdispatch.nim

```diff
@@ -77,37 +89,102 @@ proc connect*(self: RpcClient, address: string, port: Port) {.async.} =
+
+proc toJsonArray(parameters: NimNode): NimNode =
+  # outputs an array of jsonified parameters
+  # ie; %[%a, %b, %c]
+  parameters.expectKind nnkFormalParams
+  var items = newNimNode(nnkBracket)
+  for i in 2 ..< parameters.len:
+    let curParam = parameters[i][0]
+    if curParam.kind != nnkEmpty:
+      items.add(nnkPrefix.newTree(ident"%", curParam))
+  result = nnkPrefix.newTree(newIdentNode("%"), items)
+
+proc createRpcFromSig*(rpcDecl: NimNode): NimNode =
+  # Each input parameter in the rpc signature is converted
+  # to json with `%`.
+  # Return types are then converted back to native Nim types.
+  let iJsonNode = newIdentNode("JsonNode")
+
+  var parameters = rpcDecl.findChild(it.kind == nnkFormalParams).copy
+  # ensure we have at least space for a return parameter  
+  if parameters.isNil or parameters.kind == nnkEmpty or parameters.len == 0:
+    parameters = nnkFormalParams.newTree(iJsonNode)
+
+  let
+    procName = rpcDecl.name
+    pathStr = $procName
+    returnType =
+      # if no return type specified, defaults to JsonNode
+      if parameters[0].kind == nnkEmpty: iJsonNode
+      else: parameters[0]
+    customReturnType = returnType != iJsonNode
+
+  # insert rpc client as first parameter
+  parameters.insert(1, nnkIdentDefs.newTree(ident"client", ident"RpcClient", newEmptyNode()))
+
+  let
+    jsonParamIdent = genSym(nskVar, "jsonParam")  # variable used to send json to the server
+    jsonParamArray = parameters.toJsonArray()     # json array of marshalled parameters
+  var
+    # populate json params - even rpcs with no parameters have an empty json array node sent
+    callBody = newStmtList().add(quote do:
+      var `jsonParamIdent` = `jsonParamArray`
+    )
+
+  # convert return type to Future
+  parameters[0] = nnkBracketExpr.newTree(ident"Future", returnType)
+  # create rpc proc
+  result = createRpcProc(procName, parameters, callBody)
+
+  let
+    rpcResult = genSym(nskLet, "res") # temporary variable to hold `Response` from rpc call
+    procRes = ident"result"           # proc return variable
+    jsonRpcResult =                   # actual return value, `rpcResult`.result
+      nnkDotExpr.newTree(rpcResult, newIdentNode("result"))
+  
+  # perform rpc call
+  callBody.add(quote do:
+    let `rpcResult` = await client.call(`pathStr`, `jsonParamIdent`)          # `rpcResult` is of type `Response`
+    if `rpcResult`.error: raise newException(ValueError, $`rpcResult`.result) # TODO: is raise suitable here? 
+  )
+
+  if customReturnType:
+    # marshal json to native Nim type
+    callBody.add(jsonToNim(procRes, returnType, jsonRpcResult, "result"))
+  else:
+    # native json expected so no work
+    callBody.add(quote do:
+      `procRes` = `rpcResult`.result
+      )
+  when defined(nimDumpRpcs):
+    echo pathStr, ":\n", result.repr
+
+from os import getCurrentDir, DirSep
+from strutils import rsplit
+
+macro processRpcSigs(): untyped =
```

Comment:
If this is going to become a library, this macro can take a file path in the form of identifiers:

```nim
jsonRcpProcotol ../my_protocol
```
Then the NimNode tree of operators and identifiers can be converted to a path with `$`.

Reply:
Just updated my local branch to do just that :)